### ISSUE TITLE
[OJS] Allow journal managers to mark individual component types "required"

### ISSUE THREAD

**Issue by jmacgreg** [pkp/pkp-lib#3100]
For example, some journals may want to require a submission file to be uploaded during the submission process before the author can proceed with the submission. This should be an option at the "component" level.

**Comment by asmecher** [comment]
@jmacgreg, I'd suspect that journals might want to make this configuration section-specific.

**Comment by larsvilhuber** [comment]
👍

**Comment by yoruyenturk** [comment]
Hi all,

I wrote that this topic. Lots of journals want to components required and optional.

It can be added as an option in this path (Workflow Settings > Add a Component)

My forum topic:
https://forum.pkp.sfu.ca/t/ojs-submission-article-component-file-upload-required/54179

Regards,

**Comment by yoruyenturk** [comment]
Hi @asmecher @NateWr 
We realy need to require file upload to type article component. Lots of author forget some compenent during submission.
How can i change code?
Regards,

**Comment by NateWr** [comment]
Hi @yoruyenturk, you can write a generic plugin to do this. You would hook into the step 2 form's validation process and set an error if the files you expected were not found. Here's some code that would do that:

```
HookRegistry::register('submissionsubmitstep2form::validate', function($hookName, $params) {
	$form = $params[0];
	$files = DAORegistry::getDAO('SubmissionFileDAO')->getBySubmissionId($form->submissionId);
	if (/** missing files you expect */) {
		$form->addError('missingFileComponent', 'You are missing a required file component');
	}
	return false;
});
```

You can learn more about writing a plugin in our [plugin guide](https://docs.pkp.sfu.ca/dev/plugin-guide/en/). (You'll want to use a `generic` plugin.)

**Comment by yoruyenturk** [comment]
Hi @NateWr 

Thank your for your response. However, I unfortunately do not have required coding qualifications . But i'd like to make an application to PKP Sponsored Development for this issue.

**Comment by firatakin** [comment]
> Hi @yoruyenturk, you can write a generic plugin to do this. You would hook into the step 2 form's validation process and set an error if the files you expected were not found. Here's some code that would do that:
> 
> ```
> HookRegistry::register('submissionsubmitstep2form::validate', function($hookName, $params) {
> 	$form = $params[0];
> 	$files = DAORegistry::getDAO('SubmissionFileDAO')->getBySubmissionId($form->submissionId);
> 	if (/* missing files you expect */) {
> 		$form->addError('missingFileComponent', 'You are missing a required file component');
> 	}
> 	return false;
> });
> ```
> 
> You can learn more about writing a plugin in our [plugin guide](https://docs.pkp.sfu.ca/dev/plugin-guide/en/). (You'll want to use a `generic` plugin.)

Hi @NateWr 

How can I find the genreID of the files uploaded in submissionsubmitstep2form?

I'd like to write the command that "If the genreID of the uploaded files is not 1 and 2", and the condition is true for this clause how can i prevent the system from the completion of step 2?

I found something like this: https://pkp.sfu.ca/ojs/doxygen/master/html/classSubmissionFile.html#a1c3e74540a0c726414ba5a46496b3445
but I don't know right syntax.

**Comment by NateWr** [comment]
@firatakin I _think_ it would be something like this:

```php
$files = DAORegistry::getDAO('SubmissionFileDAO')->getBySubmissionId($form->submissionId);
foreach ($files as $file) {
	if ($file->getGenreId() === 1) {
		...
	}
}
```

**Comment by firatakin** [comment]
Hi @NateWr 

Thank you very much for your quick reply,
I developed a plugin on my github page. It is available at:   https://github.com/firatakin/checkSubmissionFiles 

In the step2,  i uploaded a file by describing its genreID as 2. Then i completed the step and the system did not provide any error. Could you please inform me about what is wrong in this code?

Screen record: https://nimbusweb.me/s/share/4141329/kufrvux2n1xnnhu8spmn

**Comment by NateWr** [comment]
@firatakin have you tried adding any debug code to determine what `$files` looks like when that code runs? If you're running OJS from the git repository, you can use the `Psy\sh()` tool to add breakpoints and inspect variables at runtime. Otherwise you can just try good old `print_r()`.

**Comment by asmecher** [comment]
(@firatakin, note that the genre IDs could change; it's better to rely on the genre key instead.)

**Comment by NateWr** [comment]
Closing this in favor of #6241 since that issue contains a few more details about the use cases.

**Comment by jedi-tor** [comment]
Hi @NateWr 

I am getting the following error on the latest version (3.3) of the OJS.

`Stack trace:
#0 C:\Users\*\laragon\www\ojs\lib\pkp\classes\plugins\HookRegistry.inc.php(107): checkSubmissionFilesPlugin->{closure}('submissionsubmi...', Array)
#1 C:\Users\*\laragon\www\ojs\lib\pkp\classes\form\Form.inc.php(279): HookRegistry::call('submissionsubmi...', Array)
#2 C:\Users\*\laragon\www\ojs\lib\pkp\classes\submission\form\PKPSubmissionSubmitStep2Form.inc.php(156): Form->validate(true)
#3 C:\Users\*\laragon\www\ojs\lib\pkp\pages\submission\PKPSubmissionHandler.inc.php(176): PKPSubmissionSubmitStep2Form->validate()
#4 C:\Users\*\laragon\www\ojs\lib\pkp\classes\core\PKPRouter.inc.php(395): PKPSubmissionHandler->saveStep(Array, Object(Request))
#5 C:\Users\*\laragon\www\ojs\lib\pkp\classes\core\PKPPageRouter.inc.php(246): PKPRouter->_authorizeInitializeAndCallRequest(Array, Object(Reques in C:\Users\*\laragon\www\ojs\plugins\generic\checkSubmissionFiles\checkSubmissionFilesPlugin.inc.php on line 13`


How can I fix  [this plugin](https://github.com/jedi-tor/checkSubmissionFiles/blob/master/checkSubmissionFilesPlugin.inc.php) code on line 13?
`$files = DAORegistry::getDAO('SubmissionFileDAO')->getBySubmissionId($form->submissionId);`

**Comment by NateWr** [comment]
```php
$files = Servies::get('submissionFile')->getMany(['submissionIds' => $form->submissionId]);
```

Also, `$file->getGenreId()` is deprecated. Use `$file->getData('genreId')`.

@jedi-tor in the future please post requests like this on our [support forum](https://forum.pkp.sfu.ca/).

**Comment by jedi-tor** [comment]
Hi @NateWr 

Thank you. I update the code. I add submissionId data in square brackets. Worked fine.

`$files = Services::get('submissionFile')->getMany(['submissionIds' => [$form->submissionId]]);`

**Comment by JohnKlark** [comment]
Hi,I uploaded the plugin created by @jedi-tor:

https://github.com/firatakin/checkSubmissionFiles[https://github.com/firatakin/checkSubmissionFiles](https://github.com/firatakin/checkSubmissionFiles) #
but it keeps giving me the error below
![img01](https://user-images.githubusercontent.com/100204705/155127983-132b8476-8ffb-4bab-b016-96dd8cc576de.jpg)

Should I update any information within the PHP file?
We require authors to upload two files as mandatory for the submission.

**Comment by NateWr** [comment]
@JohnKlark please report the issue on that repo. PKP does not maintain this plugin. You can report issues for that repo here:

https://github.com/jedi-tor/checkSubmissionFiles/issues

**Comment by JohnKlark** [comment]
Thank you NateWr. As you were following the previous post, could you please check if something need to be added to the php file, should I change the **genreIDs**, or the  **$reqFiles = ['1','2','3']**; for example!

**Comment by mypapit** [comment]
> Hi @NateWr
> 
> Thank you very much for your quick reply, I developed a plugin on my github page. It is available at: https://github.com/firatakin/checkSubmissionFiles
> 
> In the step2, i uploaded a file by describing its genreID as 2. Then i completed the step and the system did not provide any error. Could you please inform me about what is wrong in this code?
> 
> Screen record: https://nimbusweb.me/s/share/4141329/kufrvux2n1xnnhu8spmn

sadly, the repo is no longer available

**Comment by asmecher** [comment]
@mypapit, you'd have to contact the user that maintained that repository; they're not affiliated with PKP and are unlikely to see your comment here. But the feature the plugin implemented has been built into OJS starting with 3.4.0; see [this issue](https://github.com/pkp/pkp-lib/issues/6241) for details.
